### PR TITLE
macOS Support & AppPlatform Cleanup

### PR DESCRIPTION
**Pull request by BrentDaMage**
* macOS Support
* AppPlatform_sdlbase & AppPlatform_emscripten
* All SDL platforms still share a main.cpp, can we get this into a class?
* A shitload of other random cleanup
* Fixed Emscripten compilation + webasm build batch script

### CODE REVIEW

File: game/assets/patches/patch_data.txt
Comment (iProgramMC) [inline game/assets/patches/patch_data.txt:1]: Why did you remove this?

File: game/assets/readme.txt
Comment (iProgramMC) [inline game/assets/readme.txt:1]: And this.

File: platforms/sdl/AppPlatform_sdl.cpp [hunk 1]
```diff
@@ -1,29 +1,16 @@
...
 
 #include "compat/GL.hpp"
-#else
-#include <emscripten.h>
-#endif
 
 #include "client/common/Utils.hpp"
 
 AppPlatform_sdl::AppPlatform_sdl(std::string storageDir, SDL_Window *window)
+    : AppPlatform_sdlbase(storageDir, window, loadTexture("gui/default_world.png"))
```
Comment (iProgramMC) [inline platforms/sdl/AppPlatform_sdl.cpp:12]: Don't do this. Instead, load the texture in an init function or something.

No telling what works and what doesn't during a global constructor.

(if this is actually allocated using `new AppPlatform_XXX`, let me know)

Edit: nevermind, since you allocate it with new it's probably fine.

File: platforms/sdl/main.cpp [hunk 2]
```diff
@@ -131,46 +146,24 @@ static void handle_events()
...
-}
-
 // Resizing
 static void resize()
 {
-	SDL_GL_GetDrawableSize(window, &Minecraft::width, &Minecraft::height);
-
-	if (g_pApp != nullptr)
-	{
+    int drawWidth, drawHeight;
```
Comment (iProgramMC) [inline platforms/sdl/main.cpp:152]: The drawWidth, drawHeight, windowWidth and windowHeight ought to match. Otherwise, it will look blurry on retina displays on MacOS. Maybe increase the size of the GUI artificially if the screen size is over a certain width, like 2800 pixels? (4k screens have this problem as well)

File: source/client/gui/screens/StartMenuScreen.cpp [hunk 3]
```diff
@@ -499,7 +499,8 @@ void StartMenuScreen::render(int a, int b, float c)
 	glPushMatrix();
 
 	std::string splashText = getSplashString();
-	int textWidth = m_pFont->width(splashText), textHeight = m_pFont->height(splashText);
+    int textWidth = m_pFont->width(splashText);
```
Comment (iProgramMC) [inline source/client/gui/screens/StartMenuScreen.cpp:502]: Why did you change this?

File: source/client/player/input/ControllerTurnInput.cpp [hunk 4]
```diff
@@ -108,7 +108,7 @@ ITurnInput::Delta ControllerTurnInput::getTurnDelta()
 			if (fabsf(x1) > 0.0f)
 				x2 = x1 - x3;
 
-			float ndy = 0.0f, x5;
+            float x5;
```
Comment (iProgramMC) [inline source/client/player/input/ControllerTurnInput.cpp:111]: And this

File: source/client/player/input/ITurnInput.hpp [hunk 5]
```diff
@@ -14,7 +14,7 @@ class ITurnInput
 	struct Delta
 	{
 		float x, y;
-		Delta() { x = 0.0f, y = 0.0f; }
+        Delta() { x = 0.0f; y = 0.0f; }
```
Comment (iProgramMC) [inline source/client/player/input/ITurnInput.hpp:17]: And this

File: source/sound_data/PCM_click.h
Comment (iProgramMC) [inline source/sound_data/PCM_click.h:1]: Remove these.

File: source/sound_data/sounds.h
Comment (iProgramMC) [inline source/sound_data/sounds.h:1]: Remove these.

File: source/world/entity/Entity.cpp [hunk 6]
```diff
@@ -132,7 +132,7 @@ int Entity::move(float x, float y, float z)
 	bool b1, b2, b3, b4, b5, b6;
 	AABB hitold;
 
-	oldX = m_pos.x, oldZ = m_pos.z;
+    oldX = m_pos.x; oldZ = m_pos.z;
```
Comment (iProgramMC) [inline source/world/entity/Entity.cpp:135]: Why change this?

File: source/world/entity/Player.hpp [hunk 7]
```diff
@@ -8,7 +8,7 @@
 
 #pragma once
 
-#include "RakNetTypes.h"
+#include "thirdparty/raknet/RakNetTypes.h"
```
Comment (iProgramMC) [inline source/world/entity/Player.hpp:11]: This has no business being changed- just add raknet's include path to the includes in the macOS project.

File: thirdparty/raknet/Rand.cpp [hunk 8]
```diff
@@ -154,21 +154,22 @@ void seedMT( unsigned int seed, unsigned int *state, unsigned int *&next, int &l
...
+    left = N - 1; next = state + 1;
 
 	for ( s0 = state[ 0 ], s1 = state[ 1 ], j = N - M + 1; --j; s0 = s1, s1 = *p2++ )
 		* p0++ = *pM++ ^ ( mixBits( s0, s1 ) >> 1 ) ^ ( loBit( s1 ) ? K : 0U );
 
 	for ( pM = state, j = M; --j; s0 = s1, s1 = *p2++ )
 		* p0++ = *pM++ ^ ( mixBits( s0, s1 ) >> 1 ) ^ ( loBit( s1 ) ? K : 0U );
 
-	s1 = state[ 0 ], *p0 = *pM ^ ( mixBits( s0, s1 ) >> 1 ) ^ ( loBit( s1 ) ? K : 0U );
+    s1 = state[ 0 ];
```
Comment (iProgramMC) [inline thirdparty/raknet/Rand.cpp:171]: No business being changed.

File: tools/grabsounds.py
Comment (iProgramMC) [inline tools/grabsounds.py:1]: If you want to set this tool as executable please add a shebang at the top of the file specifying that the interpreter is python. Otherwise it will complain about an 'exec format error' or try to run it as a shell script or some other ungodly actions.

**Review by iProgramMC — CHANGES_REQUESTED** [review]
Quite a lot was changed that probably has no business being changed.

File: build-wasm.bat [hunk 9]
```diff
@@ -1,56 +1,56 @@
 @echo off
 
-:set the path where your web server's root is (iProgramInCpp's is at C:\gtcache -- don't ask why)
-set WEBSRVROOT=C:\gtcache
+::set the path where your web server's root is (iProgramInCpp's is at C:\gtcache -- don't ask why)
+set WEBSRVROOT=D:\Minecraft\MinecraftWebSite\mcpe
```
Comment (iProgramMC) [inline build-wasm.bat]: Change this back

File: build-wasm.bat [hunk 10]
```diff
@@ -1,56 +1,56 @@
...
 
-:set the emscripten root to your emscripten installation (iProgramInCpp's is at C:\emsdk)
+::set the emscripten root to your emscripten installation (iProgramInCpp's is at C:\emsdk)
 echo * Calling emsdk_env.bat.
-set EMSCRIPTEN_ROOT=C:\emsdk
+set EMSCRIPTEN_ROOT=D:\emsdk
```
Comment (iProgramMC) [inline build-wasm.bat]: And this

**Review by iProgramMC — CHANGES_REQUESTED** [review]

**Comment by iProgramMC** [comment]
Also where are the fixes in compat/GL.hpp? How did this actually build

File: platforms/sdl/main.cpp (hunk 2 above)
Comment (BrentDaMage) [reply platforms/sdl/main.cpp:152]: What you're describing is the exact opposite of how it actually works. The draw & window dimensions provided by SDL do not match on retina displays. The changes I've made have fixed all of the scaling issues found on retina displays. This has been tested on a 2015 MBP.

File: source/world/entity/Entity.cpp (hunk 6 above)
Comment (BrentDaMage) [reply source/world/entity/Entity.cpp:135]: Xcode was giving me warnings about it. The changes I made to it had no affect on the runtime of the program. Does it matter that I changed it?

File: source/client/player/input/ITurnInput.hpp (hunk 5 above)
Comment (BrentDaMage) [reply source/client/player/input/ITurnInput.hpp:17]: Xcode was giving me warnings about it. The changes I made to it had no affect on the runtime of the program. Does it matter that I changed it?

File: source/client/player/input/ControllerTurnInput.cpp (hunk 4 above)
Comment (BrentDaMage) [reply source/client/player/input/ControllerTurnInput.cpp:111]: The variable was completely unused.

File: game/assets/readme.txt
Comment (BrentDaMage) [reply game/assets/readme.txt:1]: Isn't this normally in there?

File: game/assets/patches/patch_data.txt
Comment (BrentDaMage) [reply game/assets/patches/patch_data.txt:1]: Not sure why it would've been removed.

File: thirdparty/raknet/Rand.cpp (hunk 8 above)
Comment (BrentDaMage) [reply thirdparty/raknet/Rand.cpp:171]: Why does it matter that I changed it? I think this improves readability, even if it's slight.

File: game/assets/patches/patch_data.txt
Comment (iProgramMC) [reply game/assets/patches/patch_data.txt:1]: You removed it.

File: tools/grabsounds.py
Comment (BrentDaMage) [reply tools/grabsounds.py:1]: I don't believe I made any modifications to this file. Are you seeing something that I'm not?

File: platforms/sdl/main.cpp (hunk 2 above)
Comment (iProgramMC) [reply platforms/sdl/main.cpp:152]: Interesting behavior. I guess shouldn't you then use the drawWidth and drawHeight for everything? (since the opengl surface is drawWidth x drawHeight in size)

File: build-wasm.bat (hunk 9 above)
Comment (BrentDaMage) [reply build-wasm.bat]: Why? It's only ever going to work for one person right off the bat. What difference does it make being your path or my path? It's intended that the executor of the script change it anyways.

**Comment by BrentDaMage** [comment]
> Also where are the fixes in compat/GL.hpp? How did this actually build

The file wasn't included in the Xcode project iirc, so it wasn't build with the macOS binary.

File: source/client/player/input/ITurnInput.hpp (hunk 5 above)
Comment (iProgramMC) [reply source/client/player/input/ITurnInput.hpp:17]: Well it distracts me from the things that you actually changed

File: thirdparty/raknet/Rand.cpp (hunk 8 above)
Comment (iProgramMC) [reply thirdparty/raknet/Rand.cpp:171]: I try to keep the RakNet code as close to the original (one that I grabbed).

**Comment by iProgramMC** [comment]
> The file wasn't included in the Xcode project iirc, so it wasn't build with the macOS binary.

But it is included in some places, isn't it?

File: platforms/sdl/main.cpp (hunk 2 above)
Comment (iProgramMC) [reply platforms/sdl/main.cpp:152]: Oh, wait, does SDL seriously send events divided by the draw scale? That's so messed up, yeah, probably makes more sense now that I look into it better.

File: thirdparty/raknet/Rand.cpp (hunk 8 above)
Comment (BrentDaMage) [reply thirdparty/raknet/Rand.cpp:171]: Oh that's RakNet. My bad. I wouldn't have changed that if I realized it was from RakNet. I'll undo that.

**Comment by BrentDaMage** [comment]
> > The file wasn't included in the Xcode project iirc, so it wasn't build with the macOS binary.
> 
> But it is included in some places, isn't it?

Evidently it's not used in any files the macOS build relies on.

File: source/client/player/input/ITurnInput.hpp (hunk 5 above)
Comment (BrentDaMage) [reply source/client/player/input/ITurnInput.hpp:17]: Would you like me to submit my misc. cleanups in a separate pull-request next time?

File: source/client/gui/screens/StartMenuScreen.cpp (hunk 3 above)
Comment (BrentDaMage) [reply source/client/gui/screens/StartMenuScreen.cpp:502]: The textHeight variable was unused so I split the single line up into two lines which allowed me to comment textHeight out. The changes I made to it had no affect on the runtime of the program.

File: platforms/sdl/main.cpp (hunk 2 above)
Comment (BrentDaMage) [reply platforms/sdl/main.cpp:152]: The OS sends events through SDL in a point-based system rather than by pixel. This is only made apparent when high DPI is used and causes a desync between the resolution and the DPI.

File: game/assets/patches/patch_data.txt
Comment (BrentDaMage) [reply game/assets/patches/patch_data.txt:1]: Ok. I am simply stating that I'm not sure how it ended up getting removed because I have no recollection of deleting it manually.

File: platforms/sdl/main.cpp (hunk 2 above)
Comment (iProgramMC) [reply platforms/sdl/main.cpp:152]: Alright, this is cleared up then.

File: source/client/player/input/ITurnInput.hpp (hunk 5 above)
Comment (iProgramMC) [reply source/client/player/input/ITurnInput.hpp:17]: Sure, please do.

File: tools/grabsounds.py
Comment (iProgramMC) [reply tools/grabsounds.py:1]: Yes. It shows me you made changes to the permissions of the file (you can do that with `chmod +x tools/grabsounds.py`

**Review by iProgramMC — APPROVED** [review]
